File: 69_Sqrt(x).js
Implement int sqrt(int x).

Compute and return the square root of x, where x is guaranteed to be a non-negative integer.

Since the return type is an integer, the decimal digits are truncated and only the integer part of the result is returned.

Example 1:

Input: 4
Output: 2
Example 2:

Input: 8
Output: 2
Explanation: The square root of 8 is 2.82842..., and since
             the decimal part is truncated, 2 is returned.

//easy

var mySqrt = function(x) {
 if (x === 0 || x === 1) return x;
 let start = 1;
 let end = Math.floor(x / 2);
 while (start <= end) {
   const mid = Math.floor((end - start) / 2 + start);
   const v1 = mid ** 2;
   const v2 = (mid + 1) ** 2;
   if (v1 === x || (v1 < x && v2 > x)) return mid;
   if (v1 < x) start = mid + 1;
   else if (v1 > x) end = mid - 1;
 }

 return null;
};

使用二分查找法实现.
范围是1~x/2

File: BB/69 Sqrt(x).java
Implement int sqrt(int x).

Compute and return the square root of x.

//367 Valid Perfect Square

// time Ologn space O1

class Solution {
    public int mySqrt(int x) {
        if (x <= 0) return 0;
        int low = 1, high = x;
        while(low <= high) {
          long mid = (high - low) / 2 + low;
          if (mid * mid == x) {
            return (int)mid;
          } else if (mid * mid < x) {
            low = (int)mid + 1;
          } else {
            high = (int)mid - 1;
          }
        }

        if (high * high < x) {
          return (int) high;
        } else {
          return (int) low;
        }
    }


    public int mySqrt(int x) {
      long res = x;
      while(res * res > x) {
        res = (res + x / res) / 2;
      }
      return (int) res;
    }
}
